zkevm-circuits/src/evm_circuit/execution.rs

```diff
@@ -320,24 +325,36 @@ impl<F: Field> ExecutionConfig<F> {
             execution_state_selector_constraints
                 .into_iter()
                 .map(move |(name, poly)| (name, q_usable.clone() * q_step.clone() * poly))
-            // TODO: Enable these after incomplete trace is no longer necessary.
-            // .chain(first_step_check)
-            // .chain(last_step_check)
+                .chain(first_step_check)
+                .chain(last_step_check)
         });
 
         meta.create_gate("q_step", |meta| {
             let q_usable = meta.query_selector(q_usable);
             let q_step_first = meta.query_selector(q_step_first);
+            let q_step_last = meta.query_selector(q_step_last);
             let q_step = meta.query_advice(q_step, Rotation::cur());
             let num_rows_left_cur = meta.query_advice(num_rows_until_next_step, Rotation::cur());
             let num_rows_left_next = meta.query_advice(num_rows_until_next_step, Rotation::next());
             let num_rows_left_inverse = meta.query_advice(num_rows_inv, Rotation::cur());
 
             let mut cb = BaseConstraintBuilder::default();
             // q_step needs to be enabled on the first row
-            cb.condition(q_step_first, |cb| {
+            cb.condition(q_step_first.clone(), |cb| {
+                cb.require_equal("q_step == 1", q_step.clone(), 1.expr());
+            });
+            // q_step needs to be enabled on the last row
+            cb.condition(q_step_last, |cb| {
                 cb.require_equal("q_step == 1", q_step.clone(), 1.expr());
             });
+            // rw_counter starts at 1
+            cb.condition(q_step_first, |cb| {
```

Comment:
Can we merge conditions here with line 343?

Reply:
Updated in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/795/commits/ffbeaf6e168e519926d18d93ae4c0db5a9540161